### ISSUE TITLE
[upstream] Problem with deserialization of tactic notations.

### ISSUE THREAD

**Issue by ejgallego** [ejgallego/coq-serapi#84]
As reported by @palmskog , we have:

```
$ sercomp --mode=sexp tests/genarg/tactic_alias.v > ta.sexp
$ compser ta.sexp
"Anomaly: Unknown tactic alias: SerComp.dup_#_2C9A39DD."
```
So indeed we are having the same issue than we have with tactic names cc: @ppedrot

I guess we will have to workaround it but it would be good to hear from upstream.

**Comment by ejgallego** [comment]
Something is very bizarre here. When serializing we have:
```
registering tactic alias: Coq.ZArith.BinInt.Z.nzinduct_#_#_186D5EF6 
registering tactic alias: Coq.ZArith.BinInt.Z.nzord_induct_#_401E6FEF 
registering tactic alias: Coq.ZArith.BinInt.Z.nzord_induct_#_#_401E6FEC 
registering tactic alias: SerComp.dup_#_2C9A39DD 
registering tactic alias: SerComp.dup_2C9A39DE 
```
Note the bizarre thing about `induct` being registered twice XD O_O, but ok dup gets it own alias.

Now when de-serializing we have tho:
```
registering tactic alias: Coq.ZArith.BinInt.Z.nzinduct_#_186D5EF5 
registering tactic alias: Coq.ZArith.BinInt.Z.nzinduct_#_#_186D5EF6 
registering tactic alias: Coq.ZArith.BinInt.Z.nzord_induct_#_401E6FEF 
registering tactic alias: Coq.ZArith.BinInt.Z.nzord_induct_#_#_401E6FEC 
registering tactic alias: CompSer.dup_#_1D017007 
compser: internal error, uncaught exception:
         "Anomaly: Unknown tactic alias: SerComp.dup_#_2C9A39DD."
```
so the trace for the modules is the same, however not for the module at hand!

So indeed the problem is that the current module differs, and that's captured in the hash!

boom...

Now if we set `-top` correctly indeed things work.

**Comment by palmskog** [comment]
So the bottom line is that one must make sure the same `-top` is set both during serialization and deserialization? I think this needs to be documented somewhere...

**Comment by ejgallego** [comment]
I am making up a patch, this is in fact much improved in Coq Master.

**Comment by ejgallego** [comment]
@palmskog #85 should be a good start to solve the problem, but it has two caveats:
- `foo.v` should be serialized to `foo.v`,
- unless we can backport #8991 , there is not reliable way to determine the correct module path from a file, the current heuristic is broken for invocations of the form `foo/bar.v` :(

I dunno how to proceed, let's see if we can get the fix backported indeed.

**Comment by ejgallego** [comment]
> I think this needs to be documented somewhere...

So in fact the "file name" Coq sees should be the same. Where to document that, I dunno, I'm open to suggestions.

**Comment by ejgallego** [comment]
> I dunno how to proceed, let's see if we can get the fix backported indeed.

That's the Coq backporting PR: https://github.com/coq/coq/pull/9111

**Comment by palmskog** [comment]
> So in fact the "file name" Coq sees should be the same. Where to document that, I dunno, I'm open to suggestions.

In other words, the "file name" is fixed for each SerAPI initialization? So, there would be no way to reliably (1) serialize a bunch of files separately and then (2) deserialize them all in the same SerAPI session? 

Still, I think that fixing some default top would still work, like `SerComp` in both `sercomp` and `compser`.

I'll try to think of something w.r.t. documentation after I try out the PR a bit.

**Comment by ejgallego** [comment]
> In other words, the "file name" is fixed for each SerAPI initialization?

Not exactly, the "module name" as Coq calls it, is currently set by "document. If you use `sertop` + #69 you could call `NewDoc` several times. Current code for sercomp/compser doesn't allow that as the current `Sertop_init` interface is fixed to process one single file, however this is generalized in #69.

The workaround now is to use indeed one process per file, but I will get #69 ready, then we could do `sercomp filel1 file2 file2 --mode=sexp_file`

> So, there would be no way to reliably (1) serialize a bunch of files separately and then (2) deserialize them all in the same SerAPI session?

Yup, once #69 is merged, and _critically_ the backport exposing the right API is done too. The thing is that you need to initialize the loadpath before computing the module name, and I moved that inside the `Stm.new_doc` call as the previous code was breaking all sort of invariants.

Do you foresee a problem with this approach? I may be missing something.

> Still, it sounds to me like a default global top would still work for that use case, like `SerComp` in both `sercomp` and `compser`.

That's what we have been doing for ages, however this is incorrect as @Skyskimmer correctly spotted in the Coq bug report. When you are building a module, you must set up its module name to what `coqc` would do.

> I'll try to think of something w.r.t. documentation after I try out the PR a bit.

Thanks!

**Comment by ejgallego** [comment]
Note @palmskog it is trivial to adapt this PR to work on top of https://github.com/coq/coq/pull/9111 so indeed if this is not working for your experiments, I suggest you compile the above PR and update this PR.

Things will correctly work then as long as the input module for `compser` is correctly named.

**Comment by ejgallego** [comment]
Here is the patch just in case:
```diff
diff --git a/sertop/sertop_init.ml b/sertop/sertop_init.ml
index 27c093a..54b8407 100644
--- a/sertop/sertop_init.ml
+++ b/sertop/sertop_init.ml
@@ -22,8 +22,6 @@ type async_flags = {
   deep_edits   : bool;
 }
 
-type interactive_top = TopPhysical of string | TopLogical of Names.DirPath.t
-
 type coq_opts = {
 
   (* callback to handle async feedback *)
@@ -39,7 +37,7 @@ type coq_opts = {
   aopts        : async_flags;
 
   (* name of the top-level module *)
-  top_name     : interactive_top;
+  top_name     : Stm.interactive_top;
 
   (* callback to load cma/cmo files *)
   ml_load      : (string -> unit) option;
@@ -53,8 +51,6 @@ type coq_opts = {
 (**************************************************************************)
 let coq_init opts =
 
-  let open Names in
-
   if opts.debug then begin
     Printexc.record_backtrace true;
     Flags.debug := true;
@@ -131,16 +127,8 @@ let coq_init opts =
 
   (* Flags.debug := true; *)
 
-  (* We need to declare a toplevel module name. *)
-  let sertop_dp = match opts.top_name with
-    | TopLogical mn -> mn
-    (* XXX Hack until coq/coq#8991 is released *)
-    | TopPhysical file ->
-      DirPath.make [Id.of_string String.(capitalize_ascii Filename.(remove_extension (basename file)))]
-  in
-
   (* Return the initial state of the STM *)
-  let ndoc = { Stm.doc_type = Stm.Interactive sertop_dp;
+  let ndoc = { Stm.doc_type = Stm.Interactive opts.top_name;
                require_libs = opts.require_libs;
                iload_path   = opts.iload_path;
                stm_options;
diff --git a/sertop/sertop_init.mli b/sertop/sertop_init.mli
index c49cdc9..901c973 100644
--- a/sertop/sertop_init.mli
+++ b/sertop/sertop_init.mli
@@ -22,9 +22,6 @@ type async_flags = {
   deep_edits   : bool;
 }
 
-(* See see coq/coq#8991 and ejgallego/coq-serapi#84 *)
-type interactive_top = TopPhysical of string | TopLogical of Names.DirPath.t
-
 type coq_opts = {
 
   (* callback to handle async feedback *)
@@ -40,7 +37,7 @@ type coq_opts = {
   aopts        : async_flags;
 
   (* name of the top-level module *)
-  top_name     : interactive_top;
+  top_name     : Stm.interactive_top;
 
   (* callback to load cma/cmo files *)
   ml_load      : (string -> unit) option;
diff --git a/sertop/sertop_sexp.ml b/sertop/sertop_sexp.ml
index 9b0215e..76bd974 100644
--- a/sertop/sertop_sexp.ml
+++ b/sertop/sertop_sexp.ml
@@ -118,7 +118,7 @@ let ser_loop ser_opts =
 
   let coq_path = ser_opts.coq_path in
   let iload_path = Sertop_init.coq_loadpath_default ~implicit:ser_opts.std_impl ~coq_path @ ser_opts.loadpath in
-  let top_name   = Sertop_init.(TopLogical Names.(DirPath.make [Id.of_string "SerTop"])) in
+  let top_name   = Stm.TopLogical Names.(DirPath.make [Id.of_string "SerTop"]) in
 
   (* Init Coq *)
   let _ = Sertop_init.(
```

**Comment by ejgallego** [comment]
So this is almost fixed now that I've fixed and merged #69 , however this is still broken in the interactive case so keeping this bug open until we either move to 8.10 or the backport lands in 8.9.

**Comment by ejgallego** [comment]
We are moving to 8.10 and this should be fine in 8.9 too.
